### PR TITLE
Clean up tinyprog for releasing

### PR DESCRIPTION
**Pull request by mithro**
* Use travis to test tinyprog under support Python versions.
 * Generate the version information from git and upload dev versions to PyPi.
 * Make tinyprog pep8 clean.
 * Adds a `--version` command line argument.

### CODE REVIEW

File: programmer/tinyprog/__main__.py [hunk 1]
```diff
@@ -326,10 +376,11 @@ def progress(info):
             fpga = TinyProg(active_port, progress)
 
             if args.program is not None:
-                print("    Programming " + str(active_port) + " with " + str(args.program))
+                print("    Programming %s with %s".format(
```
Comment (ewenmcneill) [inline programmer/tinyprog/__main__.py]: @mithro AFAICT you either want:

```
print("    Programming %s with %s" % (FOO, BAR))
```

or 

```
print("    Programming {0} with {1}".format(FOO, BAR))
```

as I don't remember `.format()` taking `%s` style args.  Python 3.6 says no:

```
    Programming %s with %s
    Programming at addr 028000
    Waking up SPI flash
```

Ewen

File: programmer/tinyprog/__main__.py [hunk 2]
```diff
@@ -342,16 +393,18 @@ def progress(info):
...
                     boot_fpga = True
                     print("    Programming at addr {:06x}".format(addr))
                     if not fpga.program_bitstream(addr, bitstream):
                         sys.exit(1)
-    
 
             # program user flash area
             if args.program_userdata is not None:
-                print("    Programming " + str(active_port) + " with " + str(args.program_userdata))
+                print("    Programming %s with %s".format(
```
Comment (ewenmcneill) [inline programmer/tinyprog/__main__.py]: @mithro Here too (mixed string formatting usage).

Ewen

File: programmer/tinyprog/__main__.py [hunk 3]
```diff
@@ -367,16 +420,17 @@ def progress(info):
...
                     boot_fpga = True
                     print("    Programming at addr {:06x}".format(addr))
                     if not fpga.program_bitstream(addr, bitstream):
                         sys.exit(1)
-    
 
             # program user image and data area
             if args.program_image is not None:
-                print("    Programming " + str(active_port) + " with " + str(args.program_image))
+                print("    Programming %s with %s".format(
```
Comment (ewenmcneill) [inline programmer/tinyprog/__main__.py]: @mithro Formatting string mixups...

Ewen

File: programmer/tinyprog/__init__.py [hunk 4]
```diff
@@ -419,16 +457,14 @@ def program_sectors(self, addr, data):
...
         self.progress("Success!")
         return True
 
-
     def boot(self):
         try:
             self.ser.write(b"\x00")
             self.ser.flush()
-        except:
+        except Exception:
```
Comment (ewenmcneill) [inline programmer/tinyprog/__init__.py]: @mithro is `writeTimeoutError` an exception?  I'm both unclear whether that's true, and also on one run I did have programming bail out due to a `writeTimeoutError` which I hadn't seen happen before.

Ewen

```

(tinyfpga) ewen@parthenon:/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer$ PYTHONPATH=. python3 /src/fpga/litex-buildenv/build/conda/bin/tinyprog --program-image /src/fpga/litex-buildenv/build/tinyfpga_bx_base_lm32.minimal//image-gateware+bios+micropython.bin

    TinyProg CLI
    ------------
    Using device id 1d50:6130
    Only one board with active bootloader, using it.

    Programming /dev/ttyACM0 with /src/fpga/litex-buildenv/build/tinyfpga_bx_base_lm32.minimal//image-gateware+bios+micropython.bin
    Programming at addr 028000
    Waking up SPI flash
    442884 bytes to program
    Programming and Verifying:   1%|        | 5.63k/443k [00:01<01:31, 4.80kB/s]
Traceback (most recent call last):
  File "/src/fpga/litex-buildenv/build/conda/bin/tinyprog", line 11, in <module>
    sys.exit(main())
  File "/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer/tinyprog/__main__.py", line 455, in main
    if not fpga.program_bitstream(addr, bitstream):
  File "/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 497, in program_bitstream
    return self.program_sectors(addr, bitstream)
  File "/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 441, in program_sectors
    max_length=256)
  File "/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 403, in write
    self._write(addr, write_data)
  File "/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 389, in _write
    self.wait_while_busy()
  File "/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 319, in wait_while_busy
    while to_int(self.read_sts()) & 1:
  File "/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 276, in read_sts
    return self.cmd(0x05, read_len=1)
  File "/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 262, in cmd
    self.ser.write(bytearray(cmd_write_string))
  File "/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer/tinyprog/__init__.py", line 94, in write
    self.ser.write(data)
  File "/src/tinyfpga/tools/lib/python3.5/site-packages/serial/serialposix.py", line 579, in write
    raise writeTimeoutError
serial.serialutil.SerialTimeoutException: Write timeout
(tinyfpga) ewen@parthenon:/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer$ 
```

Comment (ewenmcneill) [reply programmer/tinyprog/__init__.py]: I can find:

https://chromium.googlesource.com/experimental/chromium/src/+/84aa73c4dd94ac0089a164a7f8e8b1571c7a5eb9/tools/telemetry/third_party/pyserial/serial/serialutil.py#108

which suggests `writeTimeoutError` is a `SerialTimeoutException`:

https://chromium.googlesource.com/experimental/chromium/src/+/84aa73c4dd94ac0089a164a7f8e8b1571c7a5eb9/tools/telemetry/third_party/pyserial/serial/serialutil.py#104

which seems to be a `SerialError`:

https://chromium.googlesource.com/experimental/chromium/src/+/84aa73c4dd94ac0089a164a7f8e8b1571c7a5eb9/tools/telemetry/third_party/pyserial/serial/serialutil.py#100

which seems to be an `IOError`.  Which I *think* should be an `Exception`?  So I'm a bit unclear what's going on here.

But if it's only `writeTimeoutError` that's expected, maybe it'd make sense to be more specific?  `except Exception` doesn't seem much more useful than `except: pass`....

Ewen

File: programmer/tinyprog/__main__.py (hunk 1 above)
Comment (mithro) [reply programmer/tinyprog/__main__.py]: Fixed.

File: programmer/tinyprog/__main__.py (hunk 2 above)
Comment (mithro) [reply programmer/tinyprog/__main__.py]: Fixed.

File: programmer/tinyprog/__main__.py (hunk 3 above)
Comment (mithro) [reply programmer/tinyprog/__main__.py]: Fixed.

File: programmer/tinyprog/__init__.py (hunk 4 above)
Comment (mithro) [reply programmer/tinyprog/__init__.py]: We concluded this was a different path triggering this exception.

Comment (mithro) [reply programmer/tinyprog/__init__.py]: (We should probably also be only catching writeTimeoutError here too as you mention.)

File: programmer/tinyprog/__init__.py [hunk 5]
```diff
@@ -1,31 +1,50 @@
...
-        output += " ".join(["%02x" % ord(x) for x in data[i:i+16]]) + "\n"
+        output += " ".join(["%02x" % ord(x) for x in data[i:i + 16]]) + "\n"
     return output
 
+
 def to_int(value):
     try:
         return ord(value)
-    except:
+    except ValueError:
```
[the remark this answers is not in the document]
Comment (mithro) [reply programmer/tinyprog/__init__.py]: Fixed.

**Comment by ewenmcneill** [comment]
@mithro, FTR, the version now in master, with cleanups we went through in person seems to work okay.  So I think we just treat it as "merged by having same commits added to `master`".

Ewen

```
(tinyfpga) ewen@parthenon:/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer$ PYTHONPATH=. python3 /src/fpga/litex-buildenv/build/conda/bin/tinyprog --program-image /src/fpga/litex-buildenv/build/tinyfpga_bx_base_lm32.minimal//image-gateware+bios+micropython.bin

    TinyProg CLI
    ------------
    Using device id 1d50:6130
    Only one board with active bootloader, using it.

    Programming /dev/ttyACM0 with /src/fpga/litex-buildenv/build/tinyfpga_bx_base_lm32.minimal//image-gateware+bios+micropython.bin
    Programming at addr 028000
    Waking up SPI flash
    442884 bytes to program
    Programming and Verifying: 100%|█████████| 443k/443k [00:10<00:00, 40.6kB/s]
    Success!

(tinyfpga) ewen@parthenon:/src/tinyfpga/packages/TinyFPGA-Bootloader/programmer$
```

**Pull request merged**